test/changed.sh

```diff
@@ -17,139 +17,181 @@ set -o errexit
+                containers=$(kubectl get pods --show-all -o jsonpath="{.spec.containers[*].name}" --namespace "$NAMESPACE" "$pod")
+                for container in $containers; do
+                    printf -- '---Logs from container %s in pod %s:---\n' "$container" "$pod"
+                    kubectl logs --namespace "$NAMESPACE" -c "$container" "$pod" || true
+                    printf -- '---End of logs for container %s in pod %s---\n\n' "$container" "$pod"
+                done
+
+                printf '===End of details for pod %s===\n' "$pod"
+            fi
+        done
+
+        helm delete --purge "$current_release" > cleanup_log 2>&1 || true
+    fi
+}
 
-            echo "===End of details for pod $line===\n"
-        fi
-      done
+# Cleanup any namespace left over from the test
+cleanup_namespace() {
+    kubectl delete ns "$NAMESPACE" >> cleanup_log 2>&1 || true
+}
 
-      helm delete --purge ${CURRENT_RELEASE} > cleanup_log 2>&1 || true
+trap 'cleanup_release; cleanup_namespace' EXIT
+
+dosemvercompare() {
+    # Note, the trap and automatic exiting are disabled for the semver comparison
+    # because it catches its own errors. If the comparison fails exitCode is set
+    # to 1. So, trapping and exiting is re-enabled and then the exit is handled
+    trap - EXIT
+    set +e
+    semvercompare "$1"
+    trap 'cleanup_release; cleanup_namespace' EXIT
+    set -e
+    if [[ "$exitCode" == 1 ]]; then
+        exit 1
     fi
-    kubectl delete ns ${NAMESPACE} >> cleanup_log 2>&1 || true
 }
-trap cleanup EXIT
-
-function dosemvercompare {
-  # Note, the trap and automatic exiting are disabled for the semver comparison
-  # because it catches its own errors. If the comparison fails exitCode is set
-  # to 1. So, trapping and exiting is re-enabled and then the exit is handled
-  trap - EXIT
-  set +e
-  semvercompare ${1}
-  trap cleanup EXIT
-  set -e
-  if [ $exitCode == 1 ]; then
-    exit 1
-  fi
+
+test_release() {
+    values_file="${1:-}"
+    current_release="$release_name-$release_index"
+    release_index=$((release_index + 1))
+
+    if [[ -n "$values_file" ]]; then
+        echo "Testing chart with values file: $values_file..."
+        helm install --timeout 600 --name "$current_release" --namespace "$NAMESPACE" --values "$values_file" . | tee install_output
+    else
+        echo "Chart does not provide test values. Testing chart with defaults..."
+        helm install --timeout 600 --name "$current_release" --namespace "$NAMESPACE" . | tee install_output
+    fi
+
+    "$CURRENT_DIR/verify-release.sh" "$NAMESPACE"
+
+    kubectl get pods --namespace "$NAMESPACE"
+    kubectl get svc --namespace "$NAMESPACE"
+    kubectl get deployments --namespace "$NAMESPACE"
+    kubectl get endpoints --namespace "$NAMESPACE"
+
+    helm test "$current_release"
+
+    if [[ -n "${VERIFICATION_PAUSE:-}" ]]; then
+        cat install_output
+        sleep "$VERIFICATION_PAUSE"
+    fi
+
+    cleanup_release
```

Comment:
Now that `cleanup_release` is used for purposes other than non-zero exists, it's listing logs and details on both passing and failing test runs. Is that intentional?

Otherwise this looks good.

Reply:
Not exactly sure what you mean. Clean-up has to be done for each release. Have you looked at the test log?

Reply:
Ah, I think get it now. Didn't we print logs when test runs passed? I actually find it useful to always get the logs, especially for test logs.

Reply:
I'm fine either way. It's just a change I wanted to call out so we are explicit.